### PR TITLE
CSHARP-2697: Fix the issue with initial host resolving for replica set.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
CSHARP-2697: Fix the issue with initial host resolving for replica set.

CSHARP-2642: Add srv connection strings to atlas connectivity tests.

Evergreen: https://evergreen.mongodb.com/version/5d5c178e9ccd4e79cfd2a4cc

### CODE REVIEW

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs [hunk 1]
```diff
@@ -451,30 +448,6 @@ public void ProcessDnsResults_should_ignore_empty_end_points_list()
             }
         }
 
-        [Theory]
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs:454]: this test is related to the removed code from here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/51/files#diff-346021f0b586b815ec28dc10028a0a5aL544

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs [hunk 2]
```diff
@@ -332,16 +332,21 @@ public void IsValidHost_should_return_expected_result(string lookupDomainName, s
...
                 .SetupSequence(m => m.ShouldDnsMonitorStop())
-                .Returns(firstValue)
-                .Returns(secondValue);
+                .Returns(firstValue);
+            if (secondValue.HasValue)
+            {
+                setup.Returns(secondValue.Value);
+            }
             var subject = CreateSubject(cluster: mockCluster.Object);
+            subject._processDnsResultHasEverBeenCalled(processDnsResultHasEverBeenCalled);
```
Comment (jyemin) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs]: I don't love how this test is reliant on changing the internal state of the test subject.

Could we instead assert that in the case where cluster.ShouldDnsMonitorStop() returns true twice in a row, that `DnsResolver.ResolveSrvRecords` (using a mock resolver) is invoked exactly once before `DnsMonitor.Subject` returns normally?

**Review by jyemin — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs [hunk 3]
```diff
@@ -161,7 +161,7 @@ private void Monitor()
         {
             while (true)
             {
-                if (_cluster.ShouldDnsMonitorStop())
+                if (_processDnsResultHasEverBeenCalled && _cluster.ShouldDnsMonitorStop())
```
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: Do we need to update the logic in `ShouldDnsMonitorStop()` Currently it reads: 
```
 bool IDnsMonitoringCluster.ShouldDnsMonitorStop()
        {
            var clusterType = Description.Type;
            return clusterType != ClusterType.Unknown && clusterType != ClusterType.Sharded;
        }
```
Should the final line be`return clusterType != ClusterType.Unknown && clusterType != ClusterType.Sharded && clusterType!= ClusterType.ReplicaSet`

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs (hunk 1 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/MultiServerClusterTests.cs:454]: Ack

File: src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: No, a not added condition for `ReplicaSet` is expected by the spec. The only problem which we have that we need to call host resolving logic at least first time. All other attempts should follow the previous rules

Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: No, that wouldn't be correct. If the clusterType  != ReplicaSet, we _don't_ want to stop monitoring, we want to keep monitoring.   We _could_ add another conditional here, like:

      return (clusterType != ClusterType.Unknown && clusterType != ClusterType.Sharded_ && (isInitialized)

but that would end up being redundant to the extra check we have added here.

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: Gotcha, thanks for the explanation!

Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: And we don't have a convenient way to check here for whether it's been initialized already

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs]: Sorry, Jeff, I looked at the test once again and there is a problem which is related to the fact that `ShouldDnsMonitorStop` is not called first time in this condition:
    
    if (_processDnsResultHasEverBeenCalled && _cluster.ShouldDnsMonitorStop())

if `_processDnsResultHasEverBeenCalled`  is false.
So, without my change, we can only assert the second call of ShouldDnsMonitorStop

in this case, the test will validate not full expected logic in this method. Not sure that it's a good idea to change the current implementation

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs [hunk 4]
```diff
@@ -332,16 +332,21 @@ public void IsValidHost_should_return_expected_result(string lookupDomainName, s
...
 
             subject.Monitor();
```
Comment (vincentkam) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs:348]: Does this test just hang forever when it fails? Or does the xUnit time out?

Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs:348]: Scratch that, I see how the test ends now :)

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs (hunk 2 above)
Comment (vincentkam) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs]: I'm fine with either implementation: if Jeff's alternative isn't too much extra effort, go for it!

Comment (jyemin) [reply tests/MongoDB.Driver.Core.Tests/Core/Clusters/DnsMonitorTests.cs]: Yes, because of the short-circuit &&.

I think you could write a test which expects the following

1.  A single call to DnsResolver
2. A single call to ShouldDnsMonitorStop() (which returns true)
3. Monitor method exist

File: src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs (hunk 3 above)
Comment (vincentkam) [inline src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: Happy that this fix works, but was concerned why `ShouldDnsMonitorStop` was seemingly erroneously returning true.

Per off-line discussion with Dima, the cluster seems thinks it's a replicaset after the initial parsing of  `TXT` records due to the generated `clusterConnectionMode` option.
Dima discovered the following is called after TXT resolving: 
https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs#L1199

https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Clusters/ClusterFactory.cs#L43

TLDR: Bug seems harmless with this fix.

**Review by vincentkam — APPROVED** [review]
LGTM 😺 ! Thanks Dima!!!!

**Review by jyemin — APPROVED** [review]
I like the new test.  LGTM!

File: src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: >was concerned why ShouldDnsMonitorStop was seemingly erroneously returning true.

no, it's not an error, `ShouldDnsMonitorStop`, the result of this method is correct. Per the spec, a replica set cluster type should not be processed via this logic.

> the cluster seems thinks it's a replicaset after the initial parsing of TXT records

the original connection string describes `replica set`. So, this's expected behavior

Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs:164]: The semantics of the method are now more like `ShouldDnsMonitorStopAfterFirstSuccessfulAttempt`
